torchao/csrc/metal/mps/MPSStream.mm

```diff
@@ -0,0 +1,250 @@
+  if (!_commandBuffer) {
+    _commandBuffer = [MPSCommandBuffer commandBufferFromCommandQueue:_commandQueue].retain;
+  }
+
+  return _commandBuffer;
+}
+
+id<MTLComputeCommandEncoder> MPSStream::commandEncoder() {
+  if (!_commandEncoder) {
+    if (@available(iOS 13.0, *)) {
+      _commandEncoder = [commandBuffer() computeCommandEncoder].retain;
+    }
+  }
+
+  return _commandEncoder;
+}
+
+void MPSStream::synchronize(SyncType syncType) {
+  endKernelCoalescing();
+  switch(syncType) {
+    case SyncType::COMMIT:
+      commit();
+      break;
+    case SyncType::COMMIT_AND_WAIT:
+      commitAndWait();
+      break;
+    case SyncType::COMMIT_ADAPTIVE:
+      break;
+    case SyncType::COMMIT_AND_CONTINUE:
+      assert(
+        _enableCommitAndContinue == true && "CommitAndContinue is called but it is disabled globally!");
+      commitAndContinue();
+      break;
+    default:
+      assert(
+        false && "Unhandled syncType type");
+  }
+
+}
+
+bool MPSStream::commitAndContinueEnabled() {
+  return _enableCommitAndContinue;
+}
+
+void MPSStream::commitAndContinue() {
+  assert(_commandBuffer);
+  [_commandBuffer commitAndContinue];
+}
+
+void MPSStream::endKernelCoalescing() {
+  if (_commandEncoder) {
+    [_commandEncoder endEncoding];
+    [_commandEncoder release];
+    _commandEncoder = nil;
+  }
+}
+
+void MPSStream::commitAndWait() {
+  if (_prevCommandBuffer) {
+    // the previous command buffer (if exists) has already been committed,
+    // so we just wait until it's completed and then dispose it.
+    [_prevCommandBuffer waitUntilCompleted];
+    [_prevCommandBuffer release];
+    _prevCommandBuffer = nil;
+  }
+
+  if (_commandBuffer) {
+    [_commandBuffer commit];
+    [_commandBuffer waitUntilCompleted];
+    [_commandBuffer release];
+    _commandBuffer = nil;
+    // reset the accumulated resource sizes for command buffer
+    _commandBufferResourceSize = 0;
+  }
+}
+
+void MPSStream::commit() {
+  if (_enableCommitAndContinue) {
+    if (@available(iOS 13.0, *)) {
+      [commandBuffer() commitAndContinue];
+    }
+  } else {
+    flush();
+  }
+  // reset the accumulated resource sizes for command buffer
+  _commandBufferResourceSize = 0;
+}
+
+void MPSStream::flush() {
+  if (_commandBuffer) {
+    [_commandBuffer commit];
+    // if commitAndContinue is disabled (e.g., for Profiler), we keep the command
+    // buffer so we could wait on it later, if required.
+    if (!_enableCommitAndContinue) {
+      _prevCommandBuffer = _commandBuffer;
+    } else {
+      [_commandBuffer release];
+    }
+    _commandBuffer = nil;
+  }
+}
+
+void MPSStream::copy(id<MTLBuffer> srcBuffer,
+                     id<MTLBuffer> dstBuffer,
+                     size_t length,
+                     size_t srcOffset,
+                     size_t dstOffset,
+                     SyncType syncType) {
+  dispatch_sync(_serialQueue, ^() {
+    @autoreleasepool {
+      endKernelCoalescing();
+      if (@available(iOS 13.0, *)) {
```

Comment:
I'm a bit rusty with `@available` macro. Wouldn't it make it copy unavailable on MacOS? 

Also, what is the `else` path here? Just pretend it happened? (IMO it should error out)

Reply:
Yeah this is copied from ExecuTorch's MPSStream.mm. I think I'll refactor it so that it's using ExecuTorch's version directly.